File: drivers/zig/tests/scram_test.zig
// SCRAM-SHA-256 vector tests. The HMAC vector below is RFC 4231
// case 1; PBKDF2 + client-proof checks ensure the round-trip stays
// in lockstep with the server (`src/auth/scram.rs`) and the Rust
// driver (`drivers/rust/src/redwire/scram.rs`).

const std = @import("std");
const reddb = @import("reddb");
const scram = reddb.redwire.scram;

const t = std.testing;

test "RFC 4231 case 1 HMAC-SHA256" {
    // Key = 0x0b * 20, Data = "Hi There".
    var key: [20]u8 = .{0x0b} ** 20;
    const mac = scram.hmacSha256(&key, "Hi There");
    const expected = [_]u8{
        0xb0, 0x34, 0x4c, 0x61, 0xd8, 0xdb, 0x38, 0x53,
        0x5c, 0xa8, 0xaf, 0xce, 0xaf, 0x0b, 0xf1, 0x2b,
        0x88, 0x1d, 0xc2, 0x00, 0xc9, 0x83, 0x3d, 0xa7,
        0x26, 0xe9, 0x37, 0x6c, 0x2e, 0x32, 0xcf, 0xf7,
    };
    try t.expectEqualSlices(u8, &expected, &mac);
}

test "PBKDF2 determinism" {
    const a = try scram.pbkdf2Sha256("password", "salt", 1024);
    const b = try scram.pbkdf2Sha256("password", "salt", 1024);
    try t.expectEqualSlices(u8, &a, &b);
    const c = try scram.pbkdf2Sha256("different", "salt", 1024);
    try t.expect(!std.mem.eql(u8, &a, &c));
}

test "client_proof reproducibility" {
    const salt = "reddb-test";
    const iter: u32 = 4096;
    const password = "hunter2";
    const am = "client-first-bare,server-first,client-final-no-proof";

    const proof_a = try scram.clientProof(password, salt, iter, am);
    const proof_b = try scram.clientProof(password, salt, iter, am);
    try t.expectEqualSlices(u8, &proof_a, &proof_b);
    const proof_wrong = try scram.clientProof("wrong", salt, iter, am);
    try t.expect(!std.mem.eql(u8, &proof_a, &proof_wrong));
}

test "parseServerFirst extracts r/s/i" {
    const sf = try scram.parseServerFirst("r=abc+def,s=cmVkZGItcnQtc2FsdA==,i=4096");
    try t.expectEqualStrings("abc+def", sf.combined_nonce);
    try t.expectEqualStrings("cmVkZGItcnQtc2FsdA==", sf.salt_b64);
    try t.expectEqual(@as(u32, 4096), sf.iter);
}

test "parseServerFirst rejects malformed input" {
    try t.expectError(error.ScramBadServerFirst, scram.parseServerFirst("garbage"));
}

test "full round-trip vs recorded server_first" {
    // Reproduces the Rust driver's `scram_sha_256_end_to_end` shape:
    // simulate the client side computing a proof against a fixed
    // server-first message and ensure the same byte output drops
    // out — proves the codec stays in lockstep.
    const password = "hunter2";
    const cnonce = "fyko+d2lbbFgONRv9qkxdawL";
    const username = "alice";
    const salt = "reddb-rt-salt";
    const iter: u32 = 4096;

    const cfb = try std.fmt.allocPrint(t.allocator, "n={s},r={s}", .{ username, cnonce });
    defer t.allocator.free(cfb);

    const salt_b64 = try scram.b64Encode(t.allocator, salt);
    defer t.allocator.free(salt_b64);

    const server_first = try std.fmt.allocPrint(
        t.allocator,
        "r={s}snonce,s={s},i={d}",
        .{ cnonce, salt_b64, iter },
    );
    defer t.allocator.free(server_first);

    const sf = try scram.parseServerFirst(server_first);
    const decoded_salt = try scram.b64DecodeAlloc(t.allocator, sf.salt_b64);
    defer t.allocator.free(decoded_salt);

    const cfnp = try std.fmt.allocPrint(t.allocator, "c=biws,r={s}", .{sf.combined_nonce});
    defer t.allocator.free(cfnp);
    const am = try std.fmt.allocPrint(t.allocator, "{s},{s},{s}", .{ cfb, server_first, cfnp });
    defer t.allocator.free(am);

    const proof = try scram.clientProof(password, decoded_salt, sf.iter, am);

    // Re-derive client_proof manually and compare.
    const salted = try scram.pbkdf2Sha256(password, decoded_salt, sf.iter);
    const ck = scram.hmacSha256(&salted, "Client Key");
    const sk = scram.sha256(&ck);
    const sig = scram.hmacSha256(&sk, am);
    var expected: [32]u8 = undefined;
    scram.xor(&ck, &sig, &expected);
    try t.expectEqualSlices(u8, &expected, &proof);
}

test "base64 round-trip" {
    const src = "RedDB rocks!";
    const enc = try scram.b64Encode(t.allocator, src);
    defer t.allocator.free(enc);
    const dec = try scram.b64DecodeAlloc(t.allocator, enc);
    defer t.allocator.free(dec);
    try t.expectEqualStrings(src, dec);
}

File: drivers/cpp/tests/scram_test.cpp
// SCRAM-SHA-256 primitives + a recorded round-trip exchange.

#include "reddb/redwire/scram.hpp"

#include <gtest/gtest.h>

#include <array>
#include <cstring>
#include <string>
#include <vector>

using namespace reddb::redwire;

// RFC 4231 test case 1 — HMAC-SHA-256 known answer.
TEST(Scram, HmacSha256Rfc4231Case1) {
    std::array<uint8_t, 20> key;
    std::memset(key.data(), 0x0b, key.size());
    const char data[] = "Hi There";
    auto mac = scram::hmac_sha256(key.data(), key.size(),
                                  reinterpret_cast<const uint8_t*>(data), 8);
    const uint8_t expected[32] = {
        0xb0, 0x34, 0x4c, 0x61, 0xd8, 0xdb, 0x38, 0x53,
        0x5c, 0xa8, 0xaf, 0xce, 0xaf, 0x0b, 0xf1, 0x2b,
        0x88, 0x1d, 0xc2, 0x00, 0xc9, 0x83, 0x3d, 0xa7,
        0x26, 0xe9, 0x37, 0x6c, 0x2e, 0x32, 0xcf, 0xf7,
    };
    for (size_t i = 0; i < 32; ++i) {
        EXPECT_EQ(mac[i], expected[i]) << "byte " << i;
    }
}

TEST(Scram, Pbkdf2Deterministic) {
    auto a = scram::pbkdf2_sha256(reinterpret_cast<const uint8_t*>("password"), 8,
                                  reinterpret_cast<const uint8_t*>("salt"), 4, 4096);
    auto b = scram::pbkdf2_sha256(reinterpret_cast<const uint8_t*>("password"), 8,
                                  reinterpret_cast<const uint8_t*>("salt"), 4, 4096);
    EXPECT_EQ(a, b);
    auto c = scram::pbkdf2_sha256(reinterpret_cast<const uint8_t*>("different"), 9,
                                  reinterpret_cast<const uint8_t*>("salt"), 4, 4096);
    EXPECT_NE(a, c);
}

TEST(Scram, ProofRoundTripDeterministic) {
    std::vector<uint8_t> salt = {'r','e','d','d','b','-','t','e','s','t'};
    std::vector<uint8_t> auth(reinterpret_cast<const uint8_t*>("client-first-bare,server-first,client-final-no-proof"),
                              reinterpret_cast<const uint8_t*>("client-first-bare,server-first,client-final-no-proof") + 52);
    auto p1 = scram::client_proof("hunter2", salt, 4096, auth);
    auto p2 = scram::client_proof("hunter2", salt, 4096, auth);
    EXPECT_EQ(p1, p2);
    auto p3 = scram::client_proof("wrong", salt, 4096, auth);
    EXPECT_NE(p1, p3);
}

// Full SCRAM round trip — derive the verifier on the client side
// (the same way the engine derives it on creation), simulate the
// server signing the auth message, and verify the round trip.
TEST(Scram, FullExchangeAgainstRecordedServerFirst) {
    const std::string password = "correct horse";
    std::vector<uint8_t> salt = {'r','e','d','d','b','-','r','t','-','s','a','l','t'};
    const uint32_t iter = 4096;

    // Match what `src/auth/scram.rs::full_round_trip` builds.
    const std::string client_first_bare = "n=alice,r=cnonce";
    const std::string server_first = "r=cnonce+snonce,s=cmVkZGItcnQtc2FsdA==,i=4096";
    const std::string client_final_no_proof = "c=biws,r=cnonce+snonce";

    std::string am_str = client_first_bare + "," + server_first + "," + client_final_no_proof;
    std::vector<uint8_t> am(am_str.begin(), am_str.end());

    auto proof = scram::client_proof(password, salt, iter, am);

    // Server-side verification: ClientKey = proof XOR HMAC(stored_key, am)
    auto salted = scram::pbkdf2_sha256(
        reinterpret_cast<const uint8_t*>(password.data()), password.size(),
        salt.data(), salt.size(), iter);
    auto client_key = scram::hmac_sha256(salted.data(), salted.size(),
                                         reinterpret_cast<const uint8_t*>("Client Key"), 10);
    auto stored_key = scram::sha256(client_key.data(), client_key.size());
    auto signature = scram::hmac_sha256(stored_key.data(), stored_key.size(),
                                        am.data(), am.size());
    auto recovered_key = scram::xor_bytes(proof.data(), signature.data(), 32);
    auto derived_stored = scram::sha256(recovered_key.data(), recovered_key.size());
    EXPECT_TRUE(scram::constant_time_eq(derived_stored.data(), stored_key.data(), 32));

    // Server signature round trip — verifier reproduces the signature
    // the server would have sent in AuthOk.
    auto server_key = scram::hmac_sha256(salted.data(), salted.size(),
                                         reinterpret_cast<const uint8_t*>("Server Key"), 10);
    auto sig = scram::hmac_sha256(server_key.data(), server_key.size(),
                                  am.data(), am.size());
    std::vector<uint8_t> sig_v(sig.begin(), sig.end());
    EXPECT_TRUE(scram::verify_server_signature(password, salt, iter, am, sig_v));

    // Tampered signature is rejected.
    sig_v[0] ^= 0xFF;
    EXPECT_FALSE(scram::verify_server_signature(password, salt, iter, am, sig_v));
}

TEST(Scram, Base64RoundTrip) {
    const uint8_t data[] = {0xfe, 0x01, 0xab, 0xcd, 0xef, 0x10, 0x20, 0x30};
    auto b64 = scram::base64_encode(data, sizeof(data));
    auto back = scram::base64_decode(b64);
    ASSERT_EQ(back.size(), sizeof(data));
    for (size_t i = 0; i < sizeof(data); ++i) {
        EXPECT_EQ(back[i], data[i]);
    }
}

TEST(Scram, HexRoundTrip) {
    const uint8_t data[] = {0x00, 0xff, 0xa5, 0x5a};
    auto hex = scram::hex_encode(data, sizeof(data));
    EXPECT_EQ(hex, "00ffa55a");
    auto back = scram::hex_decode(hex);
    ASSERT_EQ(back.size(), sizeof(data));
    for (size_t i = 0; i < sizeof(data); ++i) {
        EXPECT_EQ(back[i], data[i]);
    }
}

TEST(Scram, ClientNonceIs24BytesBase64) {
    auto n1 = scram::make_client_nonce();
    auto n2 = scram::make_client_nonce();
    EXPECT_EQ(n1.size(), 32u); // 24 bytes → 32 chars base64 (no padding because 24 % 3 == 0)
    EXPECT_NE(n1, n2);         // overwhelmingly likely
    auto decoded = scram::base64_decode(n1);
    EXPECT_EQ(decoded.size(), 24u);
}
